"""
Ship FSM Handler - –∫–æ–Ω–µ—á–Ω—ã–π –∞–≤—Ç–æ–º–∞—Ç –¥–ª—è —É–ø—Ä–∞–≤–ª–µ–Ω–∏—è –∫–æ—Å–º–∏—á–µ—Å–∫–∏–º –∫–æ—Ä–∞–±–ª–µ–º.
–£–ø—Ä–∞–≤–ª—è–µ—Ç —Å–æ—Å—Ç–æ—è–Ω–∏—è–º–∏ –∫–æ—Ä–∞–±–ª—è: –∑–∞–≥—Ä—É–∑–∫–∞, —Ä–µ–∂–∏–º –æ–∂–∏–¥–∞–Ω–∏—è, –ø–æ–ª–µ—Ç, —Å—Ç—ã–∫–æ–≤–∫–∞, –∞–≤–∞—Ä–∏–π–Ω—ã–µ —Å–æ—Å—Ç–æ—è–Ω–∏—è.
"""

import sys
import os

# Add project root and generated to sys.path
project_root = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "..", "..")
)
generated_path = os.path.join(project_root, "generated")
sys.path.append(project_root)
sys.path.append(generated_path)

from typing import Dict, Any, Optional
from enum import Enum

try:
    from .interfaces import IFSMHandler
    from .agent_logger import logger
    from .ship_core import ShipCore
    from .ship_actuators import ShipActuatorController, PropulsionMode
except ImportError:
    # For direct execution
    import interfaces
    import agent_logger
    import ship_core
    import ship_actuators

    IFSMHandler = interfaces.IFSMHandler
    logger = agent_logger.logger
    ShipCore = ship_core.ShipCore
    ShipActuatorController = ship_actuators.ShipActuatorController
    PropulsionMode = ship_actuators.PropulsionMode

try:
    from generated.fsm_state_pb2 import FSMState, StateTransition
    from google.protobuf.timestamp_pb2 import Timestamp
except ImportError:
    # Mock classes for development
    class MockFSMState:
        def __init__(self):
            self.current_state_name = "SHIP_STARTUP"
            self.phase = "STARTUP"
            self.history = []
            self.timestamp = None

        def CopyFrom(self, other):
            self.current_state_name = other.current_state_name
            self.phase = other.phase
            self.history = list(other.history)

        class FSMPhase:
            STARTUP = "STARTUP"
            IDLE = "IDLE"
            FLIGHT = "FLIGHT"
            DOCKING = "DOCKING"
            EMERGENCY = "EMERGENCY"
            ERROR_STATE = "ERROR_STATE"

    class MockStateTransition:
        def __init__(self, from_state="", to_state="", trigger_event=""):
            self.from_state = from_state
            self.to_state = to_state
            self.trigger_event = trigger_event
            self.timestamp = None

    class MockTimestamp:
        def GetCurrentTime(self):
            pass

    FSMState = MockFSMState
    StateTransition = MockStateTransition
    Timestamp = MockTimestamp


class ShipState(Enum):
    """–°–æ—Å—Ç–æ—è–Ω–∏—è –∫–æ—Å–º–∏—á–µ—Å–∫–æ–≥–æ –∫–æ—Ä–∞–±–ª—è."""

    SHIP_STARTUP = "SHIP_STARTUP"  # –ó–∞–ø—É—Å–∫ —Å–∏—Å—Ç–µ–º –∫–æ—Ä–∞–±–ª—è
    SHIP_IDLE = "SHIP_IDLE"  # –ì–æ—Ç–æ–≤ –∫ –ø–æ–ª–µ—Ç—É, —Å–∏—Å—Ç–µ–º—ã –≤ —Ä–µ–∂–∏–º–µ –æ–∂–∏–¥–∞–Ω–∏—è
    FLIGHT_CRUISE = "FLIGHT_CRUISE"  # –ö—Ä–µ–π—Å–µ—Ä—Å–∫–∏–π –ø–æ–ª–µ—Ç
    FLIGHT_MANEUVERING = "FLIGHT_MANEUVERING"  # –ú–∞–Ω–µ–≤—Ä–∏—Ä–æ–≤–∞–Ω–∏–µ
    DOCKING_APPROACH = "DOCKING_APPROACH"  # –ü–æ–¥–ª–µ—Ç –∫ —Å—Ç–∞–Ω—Ü–∏–∏
    DOCKING_ENGAGED = "DOCKING_ENGAGED"  # –°—Ç—ã–∫–æ–≤–∫–∞ –≤—ã–ø–æ–ª–Ω–µ–Ω–∞
    EMERGENCY_STOP = "EMERGENCY_STOP"  # –ê–≤–∞—Ä–∏–π–Ω–∞—è –æ—Å—Ç–∞–Ω–æ–≤–∫–∞
    SYSTEMS_ERROR = "SYSTEMS_ERROR"  # –û—à–∏–±–∫–∞ —Å–∏—Å—Ç–µ–º –∫–æ—Ä–∞–±–ª—è


class ShipContext:
    """–ö–æ–Ω—Ç–µ–∫—Å—Ç —Å–æ—Å—Ç–æ—è–Ω–∏—è –∫–æ—Ä–∞–±–ª—è –¥–ª—è –ø—Ä–∏–Ω—è—Ç–∏—è —Ä–µ—à–µ–Ω–∏–π FSM."""

    def __init__(
        self, ship_core: ShipCore, actuator_controller: ShipActuatorController
    ):
        self.ship_core = ship_core
        self.actuator_controller = actuator_controller

    def is_ship_systems_ok(self) -> bool:
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç, –≤ –ø–æ—Ä—è–¥–∫–µ –ª–∏ –æ—Å–Ω–æ–≤–Ω—ã–µ —Å–∏—Å—Ç–µ–º—ã –∫–æ—Ä–∞–±–ª—è."""
        try:
            hull = self.ship_core.get_hull_status()
            power = self.ship_core.get_power_status()
            life_support = self.ship_core.get_life_support_status()
            computing = self.ship_core.get_computing_status()

            # –ö—Ä–∏—Ç–∏—á–Ω—ã–µ –ø—Ä–æ–≤–µ—Ä–∫–∏ –¥–ª—è –±–µ–∑–æ–ø–∞—Å–Ω–æ—Å—Ç–∏
            systems_ok = all(
                [
                    hull.integrity > 50.0,  # –ö–æ—Ä–ø—É—Å –Ω–µ –∫—Ä–∏—Ç–∏—á–Ω–æ –ø–æ–≤—Ä–µ–∂–¥–µ–Ω
                    power.reactor_output_mw > 0,  # –†–µ–∞–∫—Ç–æ—Ä —Ä–∞–±–æ—Ç–∞–µ—Ç
                    power.battery_charge_mwh > 0,  # –ï—Å—Ç—å –∞–≤–∞—Ä–∏–π–Ω–æ–µ –ø–∏—Ç–∞–Ω–∏–µ
                    18
                    <= life_support.atmosphere.get("oxygen_percent", 0)
                    <= 25,  # –ö–∏—Å–ª–æ—Ä–æ–¥ –≤ –Ω–æ—Ä–º–µ
                    life_support.atmosphere.get("co2_ppm", 0) < 5000,  # CO2 –Ω–µ –∫—Ä–∏—Ç–∏—á–µ–Ω
                    computing.qiki_core_status == "active",  # QIKI –∞–∫—Ç–∏–≤–µ–Ω
                ]
            )

            if not systems_ok:
                logger.warning(
                    "Ship systems check failed - some critical systems degraded"
                )

            return systems_ok

        except Exception as e:
            logger.error(f"Error checking ship systems: {e}")
            return False

    def has_navigation_capability(self) -> bool:
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç —Å–ø–æ—Å–æ–±–Ω–æ—Å—Ç—å –∫ –Ω–∞–≤–∏–≥–∞—Ü–∏–∏."""
        try:
            sensors = self.ship_core.get_sensor_status()
            propulsion = self.ship_core.get_propulsion_status()

            # –ù—É–∂–Ω—ã —Ä–∞–¥–∞—Ä –∏ –Ω–∞–≤–∏–≥–∞—Ü–∏–æ–Ω–Ω—ã–π –∫–æ–º–ø—å—é—Ç–µ—Ä + —Ä–∞–±–æ—Ç–∞—é—â–∏–µ –¥–≤–∏–≥–∞—Ç–µ–ª–∏
            navigation_ok = all(
                [
                    "long_range_radar" in sensors.active_sensors,
                    "navigation_computer" in sensors.active_sensors,
                    propulsion.main_drive_status in ["ready", "idle", "active"],
                    propulsion.main_drive_fuel_kg > 10,  # –ú–∏–Ω–∏–º—É–º —Ç–æ–ø–ª–∏–≤–∞
                ]
            )

            return navigation_ok

        except Exception as e:
            logger.error(f"Error checking navigation capability: {e}")
            return False

    def is_docking_target_in_range(self) -> bool:
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç, –µ—Å—Ç—å –ª–∏ —Ü–µ–ª—å –¥–ª—è —Å—Ç—ã–∫–æ–≤–∫–∏ –≤ —Ä–∞–¥–∏—É—Å–µ –¥–µ–π—Å—Ç–≤–∏—è."""
        track = self._get_best_station_track()
        if track is None:
            return False
        try:
            range_m = float(getattr(track, "range_m", 0.0) or 0.0)
        except Exception:
            return False
        if range_m <= 0.0:
            return False
        threshold_m = float(os.getenv("QIKI_DOCKING_TARGET_RANGE_M", "5000.0"))
        return range_m <= threshold_m

    def is_docking_engaged(self) -> bool:
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç, –≤—ã–ø–æ–ª–Ω–µ–Ω–∞ –ª–∏ —Å—Ç—ã–∫–æ–≤–∫–∞ (–ø–æ –¥–∞–Ω–Ω—ã–º —Å–µ–Ω—Å–æ—Ä–æ–≤/—Ä–∞–¥–∞—Ä —Ç—Ä–µ–∫–∞)."""
        track = self._get_best_station_track()
        if track is None:
            return False
        try:
            range_m = float(getattr(track, "range_m", 0.0) or 0.0)
            vr_mps = float(getattr(track, "vr_mps", 0.0) or 0.0)
        except Exception:
            return False
        if range_m <= 0.0:
            return False
        engaged_range_m = float(os.getenv("QIKI_DOCKING_ENGAGED_RANGE_M", "20.0"))
        max_abs_vr_mps = float(os.getenv("QIKI_DOCKING_MAX_ABS_VR_MPS", "0.5"))
        if range_m > engaged_range_m:
            return False
        if abs(vr_mps) > max_abs_vr_mps:
            return False
        return True

    def _get_best_station_track(self) -> Optional[Any]:
        """–í–æ–∑–≤—Ä–∞—â–∞–µ—Ç –±–ª–∏–∂–∞–π—à–∏–π —Ä–∞–¥–∞—Ä —Ç—Ä–µ–∫ —Ç–∏–ø–∞ STATION, –µ—Å–ª–∏ –¥–æ—Å—Ç—É–ø–µ–Ω."""
        try:
            from generated.radar.v1 import radar_pb2
        except Exception:
            return None

        best_track: Optional[Any] = None
        best_range_m: Optional[float] = None
        for reading in self.ship_core.iter_latest_sensor_readings():
            try:
                if not getattr(reading, "HasField", None):
                    continue
                if not reading.HasField("radar_track"):
                    continue
                track = reading.radar_track
            except Exception:
                continue
            try:
                if track.object_type != radar_pb2.ObjectType.STATION:
                    continue
                range_m = float(getattr(track, "range_m", 0.0) or 0.0)
            except Exception:
                continue
            if range_m <= 0.0:
                continue
            if best_range_m is None or range_m < best_range_m:
                best_track = track
                best_range_m = range_m
        return best_track

    def get_current_propulsion_mode(self) -> PropulsionMode:
        """–ü–æ–ª—É—á–∞–µ—Ç —Ç–µ–∫—É—â–∏–π —Ä–µ–∂–∏–º –¥–≤–∏–≥–∞—Ç–µ–ª—å–Ω–æ–π —Å–∏—Å—Ç–µ–º—ã."""
        return self.actuator_controller.current_mode


class ShipFSMHandler(IFSMHandler):
    """
    FSM Handler –¥–ª—è —É–ø—Ä–∞–≤–ª–µ–Ω–∏—è —Å–æ—Å—Ç–æ—è–Ω–∏—è–º–∏ –∫–æ—Å–º–∏—á–µ—Å–∫–æ–≥–æ –∫–æ—Ä–∞–±–ª—è.
    –£–ø—Ä–∞–≤–ª—è–µ—Ç –ø–µ—Ä–µ—Ö–æ–¥–∞–º–∏ –º–µ–∂–¥—É —Å–æ—Å—Ç–æ—è–Ω–∏—è–º–∏: –∑–∞–ø—É—Å–∫, –æ–∂–∏–¥–∞–Ω–∏–µ, –ø–æ–ª–µ—Ç, —Å—Ç—ã–∫–æ–≤–∫–∞, –∞–≤–∞—Ä–∏–π–Ω—ã–µ —Ä–µ–∂–∏–º—ã.
    """

    def __init__(
        self, ship_core: ShipCore, actuator_controller: ShipActuatorController
    ):
        self.ship_context = ShipContext(ship_core, actuator_controller)
        self.ship_core = ship_core
        self.actuator_controller = actuator_controller
        logger.info("ShipFSMHandler initialized for spacecraft operations.")

    def process_fsm_state(self, current_fsm_state: FSMState) -> FSMState:
        """–û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç —Ç–µ–∫—É—â–µ–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ FSM –∫–æ—Ä–∞–±–ª—è –∏ –æ–ø—Ä–µ–¥–µ–ª—è–µ—Ç —Å–ª–µ–¥—É—é—â–µ–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ."""
        logger.debug(
            f"Processing ship FSM state: {current_fsm_state.current_state_name}"
        )

        next_state = FSMState()
        if hasattr(next_state, "CopyFrom"):
            next_state.CopyFrom(current_fsm_state)
        else:
            # Fallback –¥–ª—è mock –∫–ª–∞—Å—Å–æ–≤
            next_state.current_state_name = current_fsm_state.current_state_name
            next_state.phase = current_fsm_state.phase
            next_state.history = list(current_fsm_state.history)

        # –ê–Ω–∞–ª–∏–∑ —Ç–µ–∫—É—â–µ–≥–æ —Å–æ—Å—Ç–æ—è–Ω–∏—è —Å–∏—Å—Ç–µ–º –∫–æ—Ä–∞–±–ª—è
        current_state = current_fsm_state.current_state_name
        systems_ok = self.ship_context.is_ship_systems_ok()
        nav_capable = self.ship_context.has_navigation_capability()
        docking_target = self.ship_context.is_docking_target_in_range()
        propulsion_mode = self.ship_context.get_current_propulsion_mode()

        # –õ–æ–≥–∏–∫–∞ –ø–µ—Ä–µ—Ö–æ–¥–æ–≤ —Å–æ—Å—Ç–æ—è–Ω–∏–π
        new_state_name = current_state
        trigger_event = ""
        new_phase = current_fsm_state.phase

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –ó–ê–ü–£–°–ö –ö–û–†–ê–ë–õ–Ø
        if current_state == ShipState.SHIP_STARTUP.value:
            if systems_ok and nav_capable:
                new_state_name = ShipState.SHIP_IDLE.value
                trigger_event = "SHIP_SYSTEMS_ONLINE"
                new_phase = FSMState.FSMPhase.IDLE
                logger.info("üöÄ Ship startup complete - all systems online")
            elif systems_ok and not nav_capable:
                new_state_name = ShipState.SHIP_IDLE.value
                trigger_event = "SHIP_SYSTEMS_PARTIAL"
                new_phase = FSMState.FSMPhase.IDLE
                logger.warning("‚ö†Ô∏è Ship startup with limited navigation capability")
            else:
                new_state_name = ShipState.SYSTEMS_ERROR.value
                trigger_event = "CRITICAL_SYSTEMS_FAILURE"
                new_phase = FSMState.FSMPhase.ERROR_STATE
                logger.error("‚ùå Ship startup failed - critical systems offline")

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –ì–û–¢–û–í–ù–û–°–¢–¨
        elif current_state == ShipState.SHIP_IDLE.value:
            if not systems_ok:
                new_state_name = ShipState.SYSTEMS_ERROR.value
                trigger_event = "SYSTEMS_DEGRADED"
                new_phase = FSMState.FSMPhase.ERROR_STATE
                logger.error("üö® Systems failure detected - entering error state")
            elif propulsion_mode == PropulsionMode.CRUISE:
                new_state_name = ShipState.FLIGHT_CRUISE.value
                trigger_event = "MAIN_DRIVE_ENGAGED"
                new_phase = FSMState.FSMPhase.FLIGHT
                logger.info("üåü Entering cruise flight mode")
            elif propulsion_mode == PropulsionMode.MANEUVERING:
                new_state_name = ShipState.FLIGHT_MANEUVERING.value
                trigger_event = "RCS_MANEUVERING_ACTIVE"
                new_phase = FSMState.FSMPhase.FLIGHT
                logger.info("üéØ Entering maneuvering mode")
            elif docking_target:
                new_state_name = ShipState.DOCKING_APPROACH.value
                trigger_event = "DOCKING_TARGET_ACQUIRED"
                new_phase = FSMState.FSMPhase.DOCKING
                logger.info("üéØ Docking target acquired - approaching")

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –ö–†–ï–ô–°–ï–†–°–ö–ò–ô –ü–û–õ–ï–¢
        elif current_state == ShipState.FLIGHT_CRUISE.value:
            if not systems_ok:
                new_state_name = ShipState.EMERGENCY_STOP.value
                trigger_event = "EMERGENCY_SYSTEMS_FAILURE"
                new_phase = FSMState.FSMPhase.EMERGENCY
                logger.error("üö® Emergency stop - systems failure during cruise")
                self._execute_emergency_stop()
            elif propulsion_mode == PropulsionMode.MANEUVERING:
                new_state_name = ShipState.FLIGHT_MANEUVERING.value
                trigger_event = "SWITCHING_TO_MANEUVERING"
                new_phase = FSMState.FSMPhase.FLIGHT
                logger.info("üéØ Switching from cruise to maneuvering")
            elif propulsion_mode == PropulsionMode.IDLE:
                new_state_name = ShipState.SHIP_IDLE.value
                trigger_event = "FLIGHT_COMPLETED"
                new_phase = FSMState.FSMPhase.IDLE
                logger.info("‚úÖ Flight completed - returning to idle")

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –ú–ê–ù–ï–í–†–ò–†–û–í–ê–ù–ò–ï
        elif current_state == ShipState.FLIGHT_MANEUVERING.value:
            if not systems_ok:
                new_state_name = ShipState.EMERGENCY_STOP.value
                trigger_event = "EMERGENCY_SYSTEMS_FAILURE"
                new_phase = FSMState.FSMPhase.EMERGENCY
                logger.error("üö® Emergency stop during maneuvering")
                self._execute_emergency_stop()
            elif propulsion_mode == PropulsionMode.CRUISE:
                new_state_name = ShipState.FLIGHT_CRUISE.value
                trigger_event = "SWITCHING_TO_CRUISE"
                new_phase = FSMState.FSMPhase.FLIGHT
                logger.info("üåü Switching from maneuvering to cruise")
            elif propulsion_mode == PropulsionMode.IDLE:
                new_state_name = ShipState.SHIP_IDLE.value
                trigger_event = "MANEUVERING_COMPLETED"
                new_phase = FSMState.FSMPhase.IDLE
                logger.info("‚úÖ Maneuvering completed")
            elif docking_target:
                new_state_name = ShipState.DOCKING_APPROACH.value
                trigger_event = "DOCKING_TARGET_IN_RANGE"
                new_phase = FSMState.FSMPhase.DOCKING
                logger.info("üéØ Docking target in range - beginning approach")

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –ü–û–î–õ–ï–¢ –ö –°–¢–´–ö–û–í–ö–ï
        elif current_state == ShipState.DOCKING_APPROACH.value:
            if not systems_ok:
                new_state_name = ShipState.EMERGENCY_STOP.value
                trigger_event = "EMERGENCY_DURING_DOCKING"
                new_phase = FSMState.FSMPhase.EMERGENCY
                logger.error("üö® Emergency during docking approach")
                self._execute_emergency_stop()
            elif self.ship_context.is_docking_engaged():
                new_state_name = ShipState.DOCKING_ENGAGED.value
                trigger_event = "DOCKING_COMPLETE"
                new_phase = FSMState.FSMPhase.DOCKING
                logger.info("‚úÖ Docking complete - engaged")
            elif not docking_target:
                new_state_name = ShipState.FLIGHT_MANEUVERING.value
                trigger_event = "DOCKING_TARGET_LOST"
                new_phase = FSMState.FSMPhase.FLIGHT
                logger.warning("‚ö†Ô∏è Docking target lost - returning to maneuvering")

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –ê–í–ê–†–ò–ô–ù–ê–Ø –û–°–¢–ê–ù–û–í–ö–ê
        elif current_state == ShipState.EMERGENCY_STOP.value:
            if systems_ok and propulsion_mode == PropulsionMode.EMERGENCY:
                new_state_name = ShipState.SHIP_IDLE.value
                trigger_event = "EMERGENCY_CLEARED"
                new_phase = FSMState.FSMPhase.IDLE
                logger.info("‚úÖ Emergency cleared - returning to normal operations")

        # –°–æ—Å—Ç–æ—è–Ω–∏–µ: –û–®–ò–ë–ö–ê –°–ò–°–¢–ï–ú
        elif current_state == ShipState.SYSTEMS_ERROR.value:
            if systems_ok:
                new_state_name = ShipState.SHIP_IDLE.value
                trigger_event = "SYSTEMS_RECOVERED"
                new_phase = FSMState.FSMPhase.IDLE
                logger.info("‚úÖ Systems recovered - returning to idle")

        # –í—ã–ø–æ–ª–Ω–µ–Ω–∏–µ –ø–µ—Ä–µ—Ö–æ–¥–∞ —Å–æ—Å—Ç–æ—è–Ω–∏—è
        if new_state_name != current_state:
            logger.info(
                f"üîÑ Ship FSM Transition: {current_state} -> {new_state_name} (Trigger: {trigger_event})"
            )

            new_transition = StateTransition(
                from_state=current_state,
                to_state=new_state_name,
                trigger_event=trigger_event,
            )
            if hasattr(new_transition, "timestamp"):
                timestamp = Timestamp()
                if hasattr(timestamp, "GetCurrentTime"):
                    timestamp.GetCurrentTime()
                new_transition.timestamp = timestamp

            next_state.current_state_name = new_state_name
            next_state.history.append(new_transition)
            next_state.phase = new_phase

        # –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –≤—Ä–µ–º–µ–Ω–Ω–æ–π –º–µ—Ç–∫–∏ —Å–æ—Å—Ç–æ—è–Ω–∏—è
        if hasattr(next_state, "timestamp"):
            timestamp = Timestamp()
            if hasattr(timestamp, "GetCurrentTime"):
                timestamp.GetCurrentTime()
            next_state.timestamp = timestamp

        logger.debug(f"Ship FSM new state: {next_state.current_state_name}")
        return next_state

    async def process_fsm_dto(self, current_fsm_state: Any) -> Any:
        raise NotImplementedError(
            "ShipFSMHandler.process_fsm_dto is not implemented. "
            "Use qiki.services.q_core_agent.core.fsm_handler.FSMHandler for the canonical FSM DTO path."
        )

    def _execute_emergency_stop(self):
        """–í—ã–ø–æ–ª–Ω—è–µ—Ç –∞–≤–∞—Ä–∏–π–Ω—É—é –æ—Å—Ç–∞–Ω–æ–≤–∫—É –≤—Å–µ—Ö —Å–∏—Å—Ç–µ–º –∫–æ—Ä–∞–±–ª—è."""
        try:
            logger.warning("üö® Executing emergency stop procedures")
            success = self.actuator_controller.emergency_stop()
            if success:
                logger.info("‚úÖ Emergency stop completed successfully")
            else:
                logger.error("‚ùå Emergency stop failed - manual intervention required")
        except Exception as e:
            logger.error(f"‚ùå Emergency stop execution failed: {e}")

    def get_ship_state_summary(self) -> Dict[str, Any]:
        """–ü–æ–ª—É—á–∞–µ—Ç –∫—Ä–∞—Ç–∫—É—é —Å–≤–æ–¥–∫—É —Å–æ—Å—Ç–æ—è–Ω–∏—è –∫–æ—Ä–∞–±–ª—è –¥–ª—è –¥–∏–∞–≥–Ω–æ—Å—Ç–∏–∫–∏."""
        try:
            systems_ok = self.ship_context.is_ship_systems_ok()
            nav_capable = self.ship_context.has_navigation_capability()
            propulsion_mode = self.ship_context.get_current_propulsion_mode()

            return {
                "systems_operational": systems_ok,
                "navigation_capable": nav_capable,
                "propulsion_mode": propulsion_mode.value,
                "ship_id": self.ship_core.get_id(),
                "ready_for_flight": systems_ok and nav_capable,
            }
        except Exception as e:
            logger.error(f"Error getting ship state summary: {e}")
            return {"error": str(e)}


# –ü—Ä–∏–º–µ—Ä –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è –∏ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è
if __name__ == "__main__":
    try:
        # –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –∫–æ—Ä–∞–±–µ–ª—å–Ω—ã—Ö —Å–∏—Å—Ç–µ–º
        q_core_agent_root = os.path.abspath(
            os.path.join(os.path.dirname(__file__), "..")
        )
        ship = ShipCore(base_path=q_core_agent_root)
        controller = ShipActuatorController(ship)

        # –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è FSM handler
        fsm_handler = ShipFSMHandler(ship, controller)

        print("=== SHIP FSM HANDLER TEST ===")
        print(f"Ship: {ship.get_id()}")
        print()

        # –°–æ–∑–¥–∞–Ω–∏–µ –Ω–∞—á–∞–ª—å–Ω–æ–≥–æ —Å–æ—Å—Ç–æ—è–Ω–∏—è
        initial_state = FSMState()
        initial_state.current_state_name = ShipState.SHIP_STARTUP.value
        initial_state.phase = FSMState.FSMPhase.STARTUP

        print(f"Initial state: {initial_state.current_state_name}")

        # –°–∏–º—É–ª—è—Ü–∏—è –Ω–µ—Å–∫–æ–ª—å–∫–∏—Ö —Ü–∏–∫–ª–æ–≤ FSM
        current_state = initial_state
        for i in range(5):
            print(f"\n--- FSM Cycle {i + 1} ---")
            next_state = fsm_handler.process_fsm_state(current_state)
            print(f"State: {next_state.current_state_name}")

            # –ü–æ–ª—É—á–µ–Ω–∏–µ —Å–≤–æ–¥–∫–∏ —Å–æ—Å—Ç–æ—è–Ω–∏—è
            summary = fsm_handler.get_ship_state_summary()
            print(f"Summary: {summary}")

            # –°–∏–º—É–ª—è—Ü–∏—è –∞–∫—Ç–∏–≤–Ω–æ—Å—Ç–∏ (–≤–∫–ª—é—á–µ–Ω–∏–µ –≥–ª–∞–≤–Ω–æ–≥–æ –¥–≤–∏–≥–∞—Ç–µ–ª—è –Ω–∞ 3-–º —Ü–∏–∫–ª–µ)
            if i == 2:
                print("Simulating main drive activation...")
                controller.set_main_drive_thrust(50.0)

            current_state = next_state

            # –ü—Ä–µ—Ä—ã–≤–∞–Ω–∏–µ –µ—Å–ª–∏ —Å–æ—Å—Ç–æ—è–Ω–∏–µ –Ω–µ –º–µ–Ω—è–µ—Ç—Å—è
            if (
                i > 0
                and current_state.current_state_name == next_state.current_state_name
            ):
                print("State stabilized.")
                break

    except Exception as e:
        print(f"Test failed: {e}")
        import traceback

        traceback.print_exc()
